pkg/monitor/agent/agent.go

```diff
@@ -130,11 +138,36 @@ func (a *Agent) Context() context.Context {
 	return a.ctx
 }
 
-// RegisterNewListener adds the new MonitorListener to the global list. It also spawns
-// a singleton goroutine to read and distribute the events. The goroutine is spawned
-// with a context derived from m.Context() and the cancelFunc is assigned to
-// perfReaderCancel. Note that cancelling m.Context() (e.g. on program shutdown)
-// will also cancel the derived context.
+// startPerfReaderIfFirstLocked starts the perf reader if there are no
+// listeners (and therefore no reader) yet.
+// The goroutine is spawned with a context derived from m.Context() and the
+// cancelFunc is assigned to perfReaderCancel. Note that cancelling m.Context()
+// (e.g. on program shutdown) will also cancel the derived context.
+// Note: it is critical to hold the lock for this operation.
+func (a *Agent) startPerfReaderIfFirstLocked() {
+	if len(a.listeners)+len(a.consumers) == 0 {
+		a.perfReaderCancel() // don't leak any old readers, just in case.
+		perfEventReaderCtx, cancel := context.WithCancel(a.ctx)
+		a.perfReaderCancel = cancel
+		go a.handleEvents(perfEventReaderCtx)
+	}
+}
```

Comment:
This reads a bit weird. Why not just `startPerfReader()` and keep the `if` check at call site?

Reply:
I moved the condition into a method because it needs to be shared between `RegisterNewConsumer`,  `RegisterNewListener`, `RemoveConsumer` and `RemoveListener`. If it ever needs to change, it's is crucial that all sites are updated symmetrically, otherwise we leak the perf reader. By moving it into the method, we reduce the error surface. I do agree that it looks weird, but I do believe that it is safer that way.